mslib/mscolab/server.py

```diff
@@ -124,7 +124,7 @@ def confirm_token(token, expiration=3600):
     return email
 
 
-def initialize_managers(app):
+def _initialize_managers(app):
```

Comment:
If anything imports `initialize_managers` and calls it, that would break MSColab in the same fashion. Calling `initialize_managers` more than once is always a bug. If anyone is doing that I would rather break their usage explicitly then silently run a wrongly initialized MSColab instance.

I forgot to do the same to `setup_managers` though. I fixed that now.

Reply:
In develop this function should just be removed entirely and the complete initialization process should be replaced with Flasks app factory approach. That would bring us closer to a global-variable-free and consistent application setup.

Reply:
Or actually, that _could_ happen on stable as well, if we had a defined public API. Right now basically the entirety of mslib has to be considered as the public API for SemVer changes, which turns even simple architectural changes into breaking changes. We should consider making private parts explicitly private (starting with an underscore, as is convention).

Reply:
I could keep the old names and make it so that calling the function twice would `sys.exit` with an appropriate error message. That would still technically be a breaking change though, but anything else doesn't fix the bug.

Reply:
I am still not sure if renaming is the best option when we plan to remove them. We should deprecate them with a warning too.

That is how we designed the lib currently an edge case for SemVer.  
As a human I can live with an exception but in general I think we should not start with exceptions.

Reply:
Some more options:
- emit a warning when `setup_managers` is called a second time, with the risk that people don't read warnings and run a broken MSColab
- turn the `ChatManager`, `FileManager` and `SocketsManager` objects created in `setup_managers` into global objects instead of creating new ones on each call; making calling `setup_managers` twice safe, but pointless

Reply:
not global by global that makes much problems.

Reply:
Well, the "global" keyword isn't really the issue, global variables can be set without it. But yes, mutable global variables make problems, as is evident from this very issue: the Flask app instance is such a global object, and the mismatch between that and the not-global SocketsManager broke stuff.